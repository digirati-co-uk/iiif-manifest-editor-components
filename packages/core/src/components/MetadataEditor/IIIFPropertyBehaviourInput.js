import * as React from 'react';
import * as PropTypes from 'prop-types';
import {
  FormControl,
  FormLabel,
} from '@material-ui/core';

import IIIFBehaviours from '../fields/IIIFBehaviours/IIIFBehaviours';

const IIIFPropertyBehaviourInput = ({
  labels,
  update,
  targetEntity,
  lang,
  classes,
  type
}) => (
  <FormControl component="fieldset">
    <FormLabel component="legend">
      {labels[`${type}.Behaviors`] || 'Behaviors'}
    </FormLabel>
    <IIIFBehaviours
      classes={classes}
      target={targetEntity}
      lang={lang}
      update={update}
      labels={labels}
    />
  </FormControl>
)

IIIFPropertyBehaviourInput.propTypes ={
  labels: PropTypes.object,
  update: PropTypes.func,
  targetEntity: PropTypes.object,
  lang: PropTypes.string,
  classes: PropTypes.any,
  type: PropTypes.string,
};

export default IIIFPropertyBehaviourInput;