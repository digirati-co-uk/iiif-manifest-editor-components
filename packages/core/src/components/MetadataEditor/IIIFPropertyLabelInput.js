import * as React from 'react';
import * as PropTypes from 'prop-types';

import { locale } from '../../utils/IIIFResource';
import IIIFInputField from '../fields/IIIFInputField/IIIFInputField';

const IIIFPropertyLabelInput = ({
  labels,
  update,
  targetEntity,
  lang,
  classes,
  type
}) => (
  <IIIFInputField
    label={labels[type + '.Label'] || 'Label'}
    value={locale(targetEntity.label, lang)}
    onChange={ev =>
      update(
        targetEntity,
        'label',
        lang,
        ev.target.value
      )
    }
    className={classes.textField}
  />
);

IIIFPropertyLabelInput.propTypes ={
  labels: PropTypes.object,
  update: PropTypes.func,
  targetEntity: PropTypes.object,
  lang: PropTypes.string,
  classes: PropTypes.any,
  type: PropTypes.string,
};

export default IIIFPropertyLabelInput;