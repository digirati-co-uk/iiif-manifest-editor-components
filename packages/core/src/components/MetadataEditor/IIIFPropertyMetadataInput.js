import * as React from 'react';
import * as PropTypes from 'prop-types';
import {
  FormControl,
  FormLabel,
} from '@material-ui/core';

import { locale } from '../../utils/IIIFResource';
import IIIFKeyValueField from '../fields/IIIFKeyValueField/IIIFKeyValueField';

const IIIFPropertyMetadataInput = ({
  labels,
  update,
  targetEntity,
  lang,
  type
}) => (
  <FormControl component="fieldset">
    <FormLabel component="legend">
      {labels[type + '.Metadata'] || 'Metadata'}
    </FormLabel>
    {(targetEntity.metadata || [])
      .concat({
        label: {
          [lang]: '',
        },
        value: {
          [lang]: '',
        },
      })
      .map((metadata, index) => (
        <IIIFKeyValueField
          key={`${targetEntity.id}_metadata_row__${index}:${lang}_input`}
          keyProps={{
            label:
              labels[type + '.Metadata.Label'] || 'Label',
            value: locale(metadata.label, lang),
            onChange: ev => update(
                targetEntity,
                `metadata.${index}.label`,
                lang,
                ev.target.value
              ),
          }}
          valueProps={{
            label:
              labels[type + '.Metadata.Value'] || 'Value',
            value: locale(metadata.value, lang),
            onChange: ev => update(
                targetEntity,
                `metadata.${index}.value`,
                lang,
                ev.target.value
              ),
          }}
        />
      ))}
  </FormControl>
);

IIIFPropertyMetadataInput.propTypes ={
  labels: PropTypes.object,
  update: PropTypes.func,
  targetEntity: PropTypes.object,
  lang: PropTypes.string,
  type: PropTypes.string,
};

export default IIIFPropertyMetadataInput;
