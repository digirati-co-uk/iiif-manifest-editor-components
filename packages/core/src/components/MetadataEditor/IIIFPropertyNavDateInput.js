import * as React from 'react';
import * as PropTypes from 'prop-types';
import IIIFInputField from '../fields/IIIFInputField/IIIFInputField';

const IIIFPropertyNavDateInput = ({
  labels,
  update,
  targetEntity,
  classes,
  type
}) => (
  <IIIFInputField
    label={labels[`${type}.NavDate`] || 'Nav Date'}
    type="datetime-local"
    className={classes.textField}
    value={targetEntity.navDate || ''}
    onChange={ev => update(
      targetEntity,
      'navDate',
      null,
      ev.target.value
    )}
    InputLabelProps={{
      shrink: true,
    }}
  />
);

IIIFPropertyNavDateInput.propTypes ={
  labels: PropTypes.object,
  update: PropTypes.func,
  targetEntity: PropTypes.object,
  classes: PropTypes.any,
  type: PropTypes.string,
};

export default IIIFPropertyNavDateInput;
