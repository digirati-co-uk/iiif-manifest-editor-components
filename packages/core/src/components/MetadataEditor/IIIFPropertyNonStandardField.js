
import * as React from 'react';
import * as PropTypes from 'prop-types';

import IIIFInputField from '../fields/IIIFInputField/IIIFInputField';

const IIIFPropertyNonStandardField = ({
  labels,
  update,
  targetEntity,
  lang,
  classes,
  keyName,
  type
}) => {
  const _keyName = keyName.startsWith('_')
  ? keyName.substr(1)
  : keyName;
  return (
    <IIIFInputField
      label={
        labels[`${type}.${_keyName}`] ||
        labels[_keyName] ||
        _keyName
      }
      className={classes.textField}
      value={targetEntity[_keyName] || ''}
      onChange={ev => update(
        targetEntity,
        _keyName,
        keyName.startsWith('_') ? lang : null,
        ev.target.value
      )}
    />
  );
};

IIIFPropertyNonStandardField.propTypes ={
  labels: PropTypes.object,
  update: PropTypes.func,
  targetEntity: PropTypes.object,
  lang: PropTypes.string,
  classes: PropTypes.any,
  keyName: PropTypes.string,
  type: PropTypes.string,
};

export default IIIFPropertyNonStandardField;