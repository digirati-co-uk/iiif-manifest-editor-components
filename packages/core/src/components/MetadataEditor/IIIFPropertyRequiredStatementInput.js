import * as React from 'react';
import * as PropTypes from 'prop-types';
import {
  FormControl,
  FormLabel,
} from '@material-ui/core';

import { locale } from '../../utils/IIIFResource';
import IIIFKeyValueField from '../fields/IIIFKeyValueField/IIIFKeyValueField';

const IIIFPropertyRequiredStatementInput = ({
  labels,
  update,
  targetEntity,
  lang,
  keyName,
  type
}) => (
  <FormControl component="fieldset">
    <FormLabel component="legend">
      {labels[type + '.RequiredStatement'] ||
        'Required Statement'}
    </FormLabel>
    <IIIFKeyValueField
      key={`${targetEntity.id}_property_${keyName}:${lang}_input`}
      keyProps={{
        label:
          labels[type + '.RequiredStatement.Label'] ||
          'Label',
        value: locale(
          targetEntity.requiredStatement &&
            targetEntity.requiredStatement.label,
          lang
        ),
        onChange: ev => update(
          targetEntity,
          'requiredStatement.label',
          lang,
          ev.target.value
        ),
      }}
      valueProps={{
        label:
          labels[
            type + 'Metadata.RequiredStatement.Value'
          ] || 'Value',
        value: locale(
          targetEntity.requiredStatement &&
            targetEntity.requiredStatement.value,
          lang
        ),
        onChange: ev => update(
          targetEntity,
          'requiredStatement.value',
          lang,
          ev.target.value
        ),
      }}
    />
  </FormControl>
);

IIIFPropertyRequiredStatementInput.propTypes = {
  labels: PropTypes.object,
  update: PropTypes.func,
  targetEntity: PropTypes.object,
  lang: PropTypes.string,
  keyName: PropTypes.string,
  type: PropTypes.string,
};

export default IIIFPropertyRequiredStatementInput;
