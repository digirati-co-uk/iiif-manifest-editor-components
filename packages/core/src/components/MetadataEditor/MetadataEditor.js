import * as React from 'react';
import * as PropTypes from 'prop-types';
import {
  withStyles,
  TextField,
  FormControl,
  FormLabel,
} from '@material-ui/core';
import { locale, update as rawUpdate } from '../../utils/IIIFResource';
import { throttle } from 'throttle-debounce';
import { Label, LabelConsumer } from '../LabelContext/LabelContext';
import IIIFTextField from '../IIIFTextField/IIIFTextField';
import IIIFKeyValueField from '../IIIFKeyValueField/IIIFKeyValueField';
import IIIFBehaviours from '../IIIFBehaviours/IIIFBehaviours';

const styles = theme => ({
  label: {
    backgorund: '#fff',
  },
});

class MetadataEditor extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      target: JSON.parse(JSON.stringify(props.target)),
      lastUpdate: new Date().getTime(),
    };
    this.throttledSyncProps = throttle(1000, false, this.syncLocalState);
  }

  static getDerivedStateFromProps(props, state) {
    const currentTime = new Date().getTime();
    if (
      props.target &&
      state.target.id &&
      props.target.id !== state.target.id
    ) {
      const internalRepr = JSON.parse(JSON.stringify(props.target));
      if (internalRepr.items) {
        delete internalRepr.items;
      }
      return {
        // Since this method fires on both props and state changes, local updates
        // to the controlled value will be ignored, because the props version
        // always overrides it. Oops!
        target: internalRepr,
        lastUpdate: currentTime,
      };
    }
    return null;
  }

  localUpdate = (target, prop, lang, val) => {
    const updated = rawUpdate(target, prop, lang, val);
    this.setState(
      {
        target: updated,
      },
      this.throttledSyncProps
    );
  };

  syncLocalState = () => {
    const currentTime = new Date().getTime();
    this.props.update(this.props.target, null, null, this.state.target);
    this.setState({
      lastUpdate: currentTime,
    });
  };

  render() {
    const {
      behaviorConfig,
      classes,
      target,
      lang,
      update,
      fieldConfig,
    } = this.props;
    const { type } = target || { type: '' };
    return (
      <LabelConsumer>
        {labels => (
          <React.Fragment>
            {fieldConfig.map(keyName => {
              if (typeof keyName === 'string') {
                switch (keyName) {
                  case 'label':
                    return (
                      <TextField
                        label={labels[type + '.Label'] || 'Label'}
                        key={`${target.id}_porperty_${keyName}`}
                        value={locale(this.state.target.label, lang)}
                        onChange={ev =>
                          this.localUpdate(
                            this.state.target,
                            'label',
                            lang,
                            ev.target.value
                          )
                        }
                        className={classes.textField}
                        margin="dense"
                        variant="outlined"
                      />
                    );
                  case 'summary':
                    return (
                      <IIIFTextField
                        key={`${target.id}_porperty_${keyName}`}
                        label={labels[type + '.Summary'] || 'Summary'}
                        value={locale(this.state.target.summary, lang)}
                        onChange={ev =>
                          this.localUpdate(
                            this.state.target,
                            'summary',
                            lang,
                            ev.target.value
                          )
                        }
                        className={classes.textField}
                      />
                    );
                  case 'requiredStatement':
                    return (
                      <FormControl
                        key={`${target.id}_porperty_${keyName}`}
                        component="fieldset"
                      >
                        <FormLabel component="legend">
                          {labels[type + '.RequiredStatement'] ||
                            'Required Statement'}
                        </FormLabel>
                        <IIIFKeyValueField
                          keyProps={{
                            label:
                              labels[type + '.RequiredStatement.Label'] ||
                              'Label',
                            value: locale(
                              this.state.target.requiredStatement &&
                                this.state.target.requiredStatement.label,
                              lang
                            ),
                            onChange: ev =>
                              this.localUpdate(
                                this.state.target,
                                'requiredStatement.label',
                                lang,
                                ev.target.value
                              ),
                          }}
                          valueProps={{
                            label:
                              labels[
                                type + 'Metadata.RequiredStatement.Value'
                              ] || 'Value',
                            value: locale(
                              this.state.target.requiredStatement &&
                                this.state.target.requiredStatement.value,
                              lang
                            ),
                            onChange: ev =>
                              this.localUpdate(
                                this.state.target,
                                'requiredStatement.value',
                                lang,
                                ev.target.value
                              ),
                          }}
                        />
                      </FormControl>
                    );
                  case 'metadata':
                    return (
                      <FormControl
                        key={`${target.id}_porperty_${keyName}`}
                        component="fieldset"
                      >
                        <FormLabel component="legend">
                          {labels[type + '.Metadata'] || 'Metadata'}
                        </FormLabel>
                        {(this.state.target.metadata || [])
                          .concat({
                            label: {
                              [lang]: '',
                            },
                            value: {
                              [lang]: '',
                            },
                          })
                          .map((metadata, index) => (
                            <IIIFKeyValueField
                              key={`metadata_row__${index}`}
                              keyProps={{
                                label:
                                  labels[type + '.Metadata.Label'] || 'Label',
                                value: locale(metadata.label, lang),
                                onChange: ev =>
                                  this.localUpdate(
                                    this.state.target,
                                    `metadata.${index}.label`,
                                    lang,
                                    ev.target.value
                                  ),
                              }}
                              valueProps={{
                                label:
                                  labels[type + '.Metadata.Value'] || 'Value',
                                value: locale(metadata.value, lang),
                                onChange: ev =>
                                  this.localUpdate(
                                    this.state.target,
                                    `metadata.${index}.value`,
                                    lang,
                                    ev.target.value
                                  ),
                              }}
                            />
                          ))}
                      </FormControl>
                    );
                  case 'navDate':
                    return (
                      <TextField
                        key={`${target.id}_porperty_${keyName}`}
                        label={labels[`${type}.NavDate`] || 'Nav Date'}
                        type="datetime-local"
                        className={classes.textField}
                        value={this.state.target.navDate || ''}
                        onChange={ev =>
                          this.localUpdate(
                            this.state.target,
                            'navDate',
                            null,
                            ev.target.value
                          )
                        }
                        margin="dense"
                        variant="outlined"
                        InputLabelProps={{
                          shrink: true,
                        }}
                      />
                    );
                  case 'rights':
                    return (
                      <TextField
                        key={`${target.id}_porperty_${keyName}`}
                        label={labels[`${type}.Rights`] || 'Rights'}
                        className={classes.textField}
                        value={this.state.target.rights || ''}
                        onChange={ev =>
                          this.localUpdate(
                            this.state.target,
                            'rights',
                            null,
                            ev.target.value
                          )
                        }
                        margin="dense"
                        variant="outlined"
                      />
                    );
                  case 'behavior':
                    return (
                      <FormControl
                        key={`${target.id}_porperty_${keyName}`}
                        component="fieldset"
                      >
                        <FormLabel component="legend">
                          {labels[`${type}.Behaviors`] || 'Behaviors'}
                        </FormLabel>
                        <IIIFBehaviours
                          config={behaviorConfig}
                          classes={classes}
                          target={target}
                          lang={lang}
                          update={update}
                        />
                      </FormControl>
                    );
                  default:
                    const _keyName = keyName.startsWith('_')
                      ? keyName.substr(1)
                      : keyName;
                    return (
                      <TextField
                        label={
                          labels[`${type}.${_keyName}`] ||
                          labels[_keyName] ||
                          _keyName
                        }
                        key={`${target.id}_porperty_${_keyName}`}
                        className={classes.textField}
                        value={this.state.target[_keyName] || ''}
                        onChange={ev =>
                          this.localUpdate(
                            this.state.target,
                            _keyName,
                            keyName.startsWith('_') ? lang : null,
                            ev.target.value
                          )
                        }
                        margin="dense"
                        variant="outlined"
                      />
                    );
                }
              }
            })}
          </React.Fragment>
        )}
      </LabelConsumer>
    );
  }
}

MetadataEditor.propTypes = {
  target: PropTypes.any,
  /** Language */
  lang: PropTypes.string,
  update: PropTypes.func,
  behaviorConfig: PropTypes.object,
  fieldConfig: PropTypes.array,
};

MetadataEditor.defaultProps = {
  lang: 'en',
  update: () => () => {},
  fieldConfig: null,
};

export default withStyles(styles)(MetadataEditor);
