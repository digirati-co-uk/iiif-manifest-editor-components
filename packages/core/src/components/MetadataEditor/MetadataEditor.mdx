---
name: Metadata Editor
menu: Components
---

import { Playground, PropsTable } from 'docz';
import MetadataEditor from './MetadataEditor'; 

# Metadata Editor

<PropsTable of={MetadataEditor} />