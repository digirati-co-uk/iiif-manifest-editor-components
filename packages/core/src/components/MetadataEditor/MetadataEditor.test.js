import React from 'react';
import MetadataEditor from './MetadataEditor';

describe('MetadataEditor', () => {
  it('renders the editable canvas unchanged', () => {
    const TEST_CANVAS = {
      id: 'test-canvas-1',
    };
    const TEST_ANNOTATION = {
      id: 'test-annotation-1',
    };
    const TEST_MANIFEST = {
      id: 'test-manifest-1',
    };

    const editableCanvas = shallow(
      <MetadataEditor
        manifest={TEST_MANIFEST}
        canvas={TEST_CANVAS}
        annotation={TEST_ANNOTATION}
        lang={'en'}
      />
    );
    expect(editableCanvas).toMatchSnapshot();
  });
});
