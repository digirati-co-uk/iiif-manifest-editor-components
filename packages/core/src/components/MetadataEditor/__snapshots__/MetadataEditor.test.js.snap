// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetadataEditor renders the editable canvas unchanged 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(MetadataEditor)
    annotation={
      Object {
        "id": "test-annotation-1",
      }
    }
    canvas={
      Object {
        "id": "test-canvas-1",
      }
    }
    lang="en"
    manifest={
      Object {
        "id": "test-manifest-1",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "annotation": Object {
        "id": "test-annotation-1",
      },
      "canvas": Object {
        "id": "test-canvas-1",
      },
      "classes": Object {
        "label": "MetadataEditor-label-1",
        "metadataRow": "MetadataEditor-metadataRow-2",
      },
      "lang": "en",
      "manifest": Object {
        "id": "test-manifest-1",
      },
      "update": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "annotation": Object {
          "id": "test-annotation-1",
        },
        "canvas": Object {
          "id": "test-canvas-1",
        },
        "classes": Object {
          "label": "MetadataEditor-label-1",
          "metadataRow": "MetadataEditor-metadataRow-2",
        },
        "lang": "en",
        "manifest": Object {
          "id": "test-manifest-1",
        },
        "update": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
